zebrad/src/components/inbound.rs

```diff
@@ -110,96 +131,131 @@ impl Service<zn::Request> for Inbound {
         Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
 
     fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        use oneshot::error::TryRecvError;
+
         // Check whether the network setup is finished, but don't wait for it to
         // become ready before reporting readiness. We expect to get it "soon",
         // and reporting unreadiness might cause unwanted load-shedding, since
         // the load-shed middleware is unable to distinguish being unready due
         // to load from being unready while waiting on setup.
-        if let Some(mut rx) = self.network_setup.take() {
-            use oneshot::error::TryRecvError;
-            match rx.try_recv() {
-                Ok((outbound, address_book)) => {
-                    self.outbound = Some(outbound.clone());
-                    self.address_book = Some(address_book);
-                    self.network_setup = None;
-                    self.downloads = Some(Box::pin(Downloads::new(
-                        Timeout::new(outbound, BLOCK_DOWNLOAD_TIMEOUT),
-                        Timeout::new(self.verifier.clone(), BLOCK_VERIFY_TIMEOUT),
-                        self.state.clone(),
-                    )));
-                }
-                Err(TryRecvError::Empty) => {
-                    self.network_setup = Some(rx);
+        if matches!(self.network, Setup::AwaitingNetwork { .. }) {
+            // Unfortunately, we can't match, swap, and destructure at the same time
+            let mut awaiting_state = Setup::FailedInit;
+            mem::swap(&mut self.network, &mut awaiting_state);
+            if let Setup::AwaitingNetwork {
+                mut network_setup,
+                verifier,
+            } = awaiting_state
```

Comment:
This is redundant based on the previous `matches` call, we should just use `if let` on line 141 instead

edit: this ties into the `mem::take` suggestion. You will probably want to write your own `take_setup` or something that will only swap out `self.network` if it is `AwaitingNetwork`